### PR TITLE
Pickleable pastro

### PR DESCRIPTION
**Pull request by EthanMarx**
@wbenoit26 Probably a little annoying to see this reversion to the class based implementation but I realized it was necessary  for a couple reasons:

1. Our functional approach was slick, but we couldn't `pickle` the fit models due to the use of local function definitions
2. In the online deployment, we won't want to refit every time we restart, so being able to pickle the fit model is necessary
3. This class based approach makes it a bit easier to implement different fitting procedures

### CODE REVIEW

File: libs/p_astro/p_astro/foreground.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    Foreground model that uses a KDE to model the detection statistic
+    """
+
+    def fit(self):
+        self.kde = gaussian_kde(self.foreground.detection_statistic)
+
+    def __call__(self, stats):
+        density = self.kde(stats)
+        if isinstance(stats, (float, int)):
+            density = density.item()
```
Comment (wbenoit26) [inline libs/p_astro/p_astro/foreground.py:106]: Does a similar check need to happen in `Background`?

**Comment by wbenoit26** [comment]
This looks great, and makes a ton of sense. I think we should remove the ability to downsample the background because doing so generally produces inaccurate results.

File: libs/p_astro/p_astro/foreground.py (hunk 1 above)
Comment (EthanMarx) [reply libs/p_astro/p_astro/foreground.py:106]: For whatever reason `np.piecewise` handles this, have this check in the tests.

**Comment by EthanMarx** [comment]
Okay great glad you agree - yeah I can remove that behavior until we figure it out - was passing `None` to that variable anyway

**Review by wbenoit26 — APPROVED** [review]

**Review by wbenoit26 — APPROVED** [review]

**Pull request merged**